File: crates/core/src/message.rs
use std::fmt;
use std::pin::Pin;

use futures_core::Stream;
use serde::{Deserialize, Serialize};
use serde_json::Value;
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Role {
    System,
    User,
    Assistant,
    Tool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub id: String,
    pub name: String,
    pub args: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "content", rename_all = "snake_case")]
pub enum Content {
    Text(String),
    ToolCall(ToolCall),
}

#[derive(Debug, Clone)]
pub struct Message {
    pub role: Role,
    pub content: Vec<Content>,
    /// Set for Role::Tool so the result can be matched to its call.
    pub tool_call_id: Option<String>,
}

impl Message {
    pub fn user(text: impl Into<String>) -> Self {
        Self { role: Role::User, content: vec![Content::Text(text.into())], tool_call_id: None }
    }
    pub fn system(text: impl Into<String>) -> Self {
        Self { role: Role::System, content: vec![Content::Text(text.into())], tool_call_id: None }
    }
    pub fn assistant_text(text: impl Into<String>) -> Self {
        Self { role: Role::Assistant, content: vec![Content::Text(text.into())], tool_call_id: None }
    }
    pub fn assistant_tool_call(call: ToolCall) -> Self {
        Self { role: Role::Assistant, content: vec![Content::ToolCall(call)], tool_call_id: None }
    }
    pub fn tool_result(id: impl Into<String>, text: impl Into<String>) -> Self {
        Self { role: Role::Tool, content: vec![Content::Text(text.into())], tool_call_id: Some(id.into()) }
    }
}

/// The model-facing view of a tool: name, description, and a JSON Schema for args.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolSpec {
    pub name: String,
    pub description: String,
    pub parameters: Value,
}

#[derive(Debug, Clone)]
pub struct ModelRequest {
    pub model: String,
    pub system: Option<String>,
    pub messages: Vec<Message>,
    pub tools: Vec<ToolSpec>,
    pub max_tokens: Option<u32>,
    pub temperature: Option<f32>,
}

/// Token usage reported by a provider for one LLM response (W220).
///
/// Mirrors the OpenAI-compatible `usage` object; providers fill a subset.
/// `cache_read` is the provider-reported cache-hit prompt tokens, whatever key
/// it arrives under (deepseek `prompt_cache_hit_tokens`, openai
/// `prompt_tokens_details.cached_tokens`, ...). Consumers accumulate it with
/// [Usage::add] / AddAssign.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct Usage {
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    pub total_tokens: u64,
    pub cache_read: u64,
    pub reasoning_tokens: u64,
}

impl Usage {
    /// True when every counter is zero — used to skip empty usage payloads.
    pub fn is_empty(&self) -> bool {
        self.prompt_tokens == 0
            && self.completion_tokens == 0
            && self.total_tokens == 0
            && self.cache_read == 0
            && self.reasoning_tokens == 0
    }

    /// Accumulate `other` into `self` (per-field sum).
    pub fn add(&mut self, other: &Usage) {
        self.prompt_tokens += other.prompt_tokens;
        self.completion_tokens += other.completion_tokens;
        self.total_tokens += other.total_tokens;
        self.cache_read += other.cache_read;
        self.reasoning_tokens += other.reasoning_tokens;
    }
}

impl std::ops::AddAssign for Usage {
    fn add_assign(&mut self, rhs: Self) {
        self.add(&rhs);
    }
}

/// Stream events from a provider. Providers emit incremental deltas for the
/// UI, then a single final Message (text + tool calls) as the last event.
///
/// - StreamEvent::Text — a final-answer text delta.
/// - StreamEvent::Thinking — a chain-of-thought / reasoning delta from a
///   reasoning model (e.g. DeepSeek reasoning_content). Carried as its own
///   event so consumers can distinguish it from the final answer (W191).
/// - StreamEvent::Usage — provider-reported token usage for this response,
///   emitted when the provider reports it (stream-end usage frame or final
///   chunk), just before the authoritative Done (W220).
/// - StreamEvent::Done — the single authoritative final message.
#[derive(Debug, Clone)]
pub enum StreamEvent {
    Text(String),
    Thinking(String),
    Usage(Usage),
    Done(Message),
}

pub type LlmStream = Pin<Box<dyn Stream<Item = StreamEvent> + Send>>;

#[derive(Debug, Clone)]
pub struct LlmError(pub String);

impl fmt::Display for LlmError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for LlmError {}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn message_constructors_shape_content() {
        let m = Message::user("hi");
        assert!(matches!(m.role, Role::User));
        let tc = Message::assistant_tool_call(ToolCall {
            id: "c1".into(),
            name: "read_file".into(),
            args: serde_json::json!({ "path": "a" }),
        });
        assert!(matches!(tc.content.as_slice(), [Content::ToolCall(_)]));
        let tr = Message::tool_result("c1", "ok");
        assert_eq!(tr.tool_call_id.as_deref(), Some("c1"));
        assert!(matches!(tr.role, Role::Tool));
    }
    #[test]
    fn stream_event_thinking_is_a_distinct_variant() {
        // Thinking carries its own payload, distinct from Text and Done.
        let thinking = StreamEvent::Thinking("reasoning...".to_string());
        assert!(matches!(&thinking, StreamEvent::Thinking(_)));
        assert!(!matches!(&thinking, StreamEvent::Text(_)));
        assert!(!matches!(&thinking, StreamEvent::Done(_)));
    }

    #[test]
    fn stream_event_usage_is_a_distinct_variant() {
        // Usage carries token counters, distinct from Text / Thinking / Done.
        let u = StreamEvent::Usage(Usage {
            prompt_tokens: 10,
            completion_tokens: 20,
            total_tokens: 30,
            cache_read: 4,
            reasoning_tokens: 3,
        });
        assert!(matches!(&u, StreamEvent::Usage(_)));
        assert!(!matches!(&u, StreamEvent::Text(_)));
        assert!(!matches!(&u, StreamEvent::Thinking(_)));
        assert!(!matches!(&u, StreamEvent::Done(_)));
    }

    #[test]
    fn usage_accumulates_and_reports_empty() {
        let mut a = Usage {
            prompt_tokens: 10,
            completion_tokens: 20,
            total_tokens: 30,
            cache_read: 4,
            reasoning_tokens: 3,
        };
        let b = Usage {
            prompt_tokens: 1,
            completion_tokens: 2,
            total_tokens: 3,
            cache_read: 1,
            reasoning_tokens: 1,
        };
        a.add(&b);
        assert_eq!(a.prompt_tokens, 11);
        assert_eq!(a.completion_tokens, 22);
        assert_eq!(a.total_tokens, 33);
        assert_eq!(a.cache_read, 5);
        assert_eq!(a.reasoning_tokens, 4);
        // AddAssign is the same as add.
        a += b;
        assert_eq!(a.prompt_tokens, 12);
        assert_eq!(a.completion_tokens, 24);
        // Empty usage is detectable (usage-only frames can carry zeroes).
        assert!(Usage::default().is_empty());
        assert!(!a.is_empty());
    }

    #[test]
    fn usage_serde_roundtrip() {
        let u = Usage {
            prompt_tokens: 100,
            completion_tokens: 200,
            total_tokens: 300,
            cache_read: 50,
            reasoning_tokens: 40,
        };
        let json = serde_json::to_string(&u).unwrap();
        assert_eq!(
            serde_json::from_str::<Usage>(&json).unwrap(),
            u,
            "usage must serialize/deserialize losslessly (flat counters)"
        );
    }

    #[test]
    fn message_system_constructor() {
        let m = Message::system("be terse");
        assert_eq!(m.role, Role::System);
        assert!(matches!(m.content.as_slice(), [Content::Text(t)] if t == "be terse"));
        assert!(m.tool_call_id.is_none());
    }

}

File: packages/core/src/message.ts
/**
 * The model-visible message model — a 1:1 port of
 * `celestea_harness/crates/core/src/message.rs`.
 *
 * Shapes ported here:
 *   Role      enum { System, User, Assistant, Tool }      (serde lowercase)
 *   ToolCall  { id, name, args: Value }
 *   Content   enum { Text(String), ToolCall(ToolCall) }   (tag="type", content="content")
 *   Message   { role, content: Vec<Content>, tool_call_id: Option<String> }
 *   ToolSpec  { name, description, parameters }           (declared in ./types.ts)
 *   Usage     { prompt_tokens, completion_tokens, total_tokens,
 *               cache_read, reasoning_tokens }             (flat counters)
 *
 * Field names and content-tag names are contract, not style: the LLM request
 * builder and every engine projection depend on them. Do not rename anything.
 */

// ---------------------------------------------------------------------------
// Role / Content / ToolCall / Message
// ---------------------------------------------------------------------------

/** `Role` — serde `rename_all = "lowercase"`. */
export const ROLES = ["system", "user", "assistant", "tool"] as const;
export type Role = (typeof ROLES)[number];

/** `ToolCall` — the id is the provider call id, `args` is the raw JSON value. */
export interface ToolCall {
  id: string;
  name: string;
  args: unknown;
}

/** `Content::Text` — `{"type":"text","content":"…"}`. */
export interface TextContent {
  type: "text";
  content: string;
}

/** `Content::ToolCall` — `{"type":"tool_call","content":{…}}`. */
export interface ToolCallContent {
  type: "tool_call";
  content: ToolCall;
}

/** The four media types the attachment pipeline accepts (section 5.4 magic-byte whitelist). */
export const IMAGE_MEDIA_TYPES = ["image/png", "image/jpeg", "image/webp", "image/gif"] as const;
export type ImageMediaType = (typeof IMAGE_MEDIA_TYPES)[number];

/** True for one of the four accepted normalized media types. */
export function isImageMediaType(v: unknown): v is ImageMediaType {
  return typeof v === "string" && (IMAGE_MEDIA_TYPES as readonly string[]).includes(v);
}

/**
 * W804 (multimodal P0 4.1): a content-addressed REFERENCE to an attachment
 * object, never the bytes. The object itself lives at
 * <session-dir>/attachments/<attachment_id>.<ext> (section 5); the session log
 * and every projected event carry only this shape — base64 in the log is a red line.
 *
 * original records the pre-normalization dimensions of an image that was
 * re-encoded/downscaled (DSH originalDimensions semantics); P0 never writes it
 * (no re-encode), but the type is frozen here so P1 is purely additive.
 */
export interface ImageRef {
  /** Content-addressing id: sha256(original bytes) in lowercase hex. */
  attachment_id: string;
  /** Normalized media type (magic-byte sniffed, never trusted from a name). */
  media_type: ImageMediaType;
  /** Normalized pixel width. */
  width: number;
  /** Normalized pixel height. */
  height: number;
  /** Original upload/file name, for the UI and model-readable labels only. */
  name?: string;
  /** Set when the stored bytes are a downscaled/re-encoded variant. */
  original?: { width: number; height: number; bytes: number; media_type: string };
}

/**
 * W804: the log/event spelling of ImageRef (section 4.3). Structurally identical —
 * the alias exists so the event codec and the content model can each use the
 * vocabulary of their own layer without drifting apart.
 */
export type AttachmentRef = ImageRef;

/** Content::Image — {"type":"image","content":{...ImageRef}} (section 4.1). */
export interface ImageContent {
  type: "image";
  content: ImageRef;
}

export type Content = TextContent | ToolCallContent | ImageContent;

/**
 * `Message` — one entry of the model-visible history.
 *
 * `content` is always a list (an assistant turn that calls tools carries one
 * `tool_call` item per call in a SINGLE message); `tool_call_id` is set only for
 * `role = "tool"`, so a result can be matched to its call.
 */
export interface Message {
  role: Role;
  content: Content[];
  tool_call_id: string | null;
}

// Constructors mirror the engine's `impl Message` (same names, camelCase).

/** `Message::user`. */
export function userMessage(text: string): Message {
  return { role: "user", content: [{ type: "text", content: text }], tool_call_id: null };
}

/** `Message::system`. */
export function systemMessage(text: string): Message {
  return { role: "system", content: [{ type: "text", content: text }], tool_call_id: null };
}

/** `Message::assistant_text`. */
export function assistantText(text: string): Message {
  return { role: "assistant", content: [{ type: "text", content: text }], tool_call_id: null };
}

/** `Message::assistant_tool_call`. */
export function assistantToolCall(call: ToolCall): Message {
  return { role: "assistant", content: [{ type: "tool_call", content: call }], tool_call_id: null };
}

/** `Message::tool_result`. */
export function toolResultMessage(id: string, text: string): Message {
  return { role: "tool", content: [{ type: "text", content: text }], tool_call_id: id };
}

/**
 * W804: a user message that carries text plus image references. The text block
 * stays FIRST and the images follow (section 3.5 ordering rule); no-image callers
 * keep using userMessage so the wire shape is byte-identical when there is
 * nothing to attach.
 */
export function userMessageWithImages(text: string, images: readonly ImageRef[]): Message {
  const content: Content[] = [{ type: "text", content: text }];
  for (const ref of images) content.push(imageContent(ref));
  return { role: "user", content, tool_call_id: null };
}

/**
 * W804: a tool result that carries the canonical JSON text plus image
 * references (the read_image case, section 6.3). The TEXT remains the value's
 * serde JSON — the model must still see path/dimensions — and the images are
 * separate content blocks.
 */
export function toolResultWithImages(id: string, text: string, images: readonly ImageRef[]): Message {
  const content: Content[] = [{ type: "text", content: text }];
  for (const ref of images) content.push(imageContent(ref));
  return { role: "tool", content, tool_call_id: id };
}

/** Static namespace facade: `Message::user(…)` → `Message.user(…)`. */
export const Message = {
  user: userMessage,
  system: systemMessage,
  assistantText,
  assistantToolCall,
  toolResult: toolResultMessage,
  userWithImages: userMessageWithImages,
  toolResultWithImages,
} as const;

// ---------------------------------------------------------------------------
// Content helpers (the engine pattern-matches; these are the TS equivalents)
// ---------------------------------------------------------------------------

export function isTextContent(c: Content): c is TextContent {
  return c.type === "text";
}

export function isToolCallContent(c: Content): c is ToolCallContent {
  return c.type === "tool_call";
}

export function isImageContent(c: Content): c is ImageContent {
  return c.type === "image";
}

/** Wrap one already-content-addressed reference as a model-visible image block. */
export function imageContent(ref: ImageRef): ImageContent {
  return { type: "image", content: ref };
}

/** The image blocks of a message, in order (references only, never bytes). */
export function messageImages(m: Message): ImageRef[] {
  const out: ImageRef[] = [];
  for (const c of m.content) if (isImageContent(c)) out.push(c.content);
  return out;
}

/** The tool calls carried by a message (empty for a text-only message). */
export function messageToolCalls(m: Message): ToolCall[] {
  const out: ToolCall[] = [];
  for (const c of m.content) if (isToolCallContent(c)) out.push(c.content);
  return out;
}

/** The ids of the tool calls carried by a message. */
export function toolCallIds(m: Message): string[] {
  return messageToolCalls(m).map((tc) => tc.id);
}

/**
 * The text blocks of a message, in order. W804: image blocks are deliberately
 * NOT text — they are never concatenated here (and never stringified into a
 * prompt); a consumer that needs the images reads messageImages.
 */
export function messageTexts(m: Message): string[] {
  const out: string[] = [];
  for (const c of m.content) if (isTextContent(c)) out.push(c.content);
  return out;
}

/** The single text block of a text-only message, or null. */
export function messageText(m: Message): string | null {
  const texts = messageTexts(m);
  return texts.length === 1 ? (texts[0] ?? null) : null;
}

/** True when the message carries at least one tool call. */
export function hasToolCalls(m: Message): boolean {
  return m.content.some(isToolCallContent);
}

// ---------------------------------------------------------------------------
// ---------------------------------------------------------------------------
// Attachment reference helpers (W804: the log/event side of ImageRef)
// ---------------------------------------------------------------------------

/**
 * True when v has the ImageRef shape the event codec and projection accept.
 * name and original are optional and their absence is the normal case.
 */
export function isImageRef(v: unknown): v is ImageRef {
  if (typeof v !== "object" || v === null) return false;
  const r = v as Record<string, unknown>;
  return (
    typeof r["attachment_id"] === "string" &&
    isImageMediaType(r["media_type"]) &&
    typeof r["width"] === "number" &&
    Number.isFinite(r["width"]) &&
    typeof r["height"] === "number" &&
    Number.isFinite(r["height"]) &&
    (r["name"] === undefined || typeof r["name"] === "string")
  );
}

/**
 * Field-whitelist normalization of one decoded attachment reference: unknown
 * fields are dropped and absent optionals stay omitted (serde style). Returns
 * null when the value is not an ImageRef at all.
 */
export function normalizeImageRef(v: unknown): ImageRef | null {
  if (!isImageRef(v)) return null;
  const out: ImageRef = {
    attachment_id: v.attachment_id,
    media_type: v.media_type,
    width: v.width,
    height: v.height,
  };
  if (typeof v.name === "string") out.name = v.name;
  if (typeof v.original === "object" && v.original !== null) {
    const o = v.original as Record<string, unknown>;
    if (
      typeof o["width"] === "number" &&
      typeof o["height"] === "number" &&
      typeof o["bytes"] === "number" &&
      typeof o["media_type"] === "string"
    ) {
      out.original = { width: o["width"], height: o["height"], bytes: o["bytes"], media_type: o["media_type"] };
    }
  }
  return out;
}

/**
 * The image references carried by a tool_result value (its attachments field,
 * section 6.3). Anything that is not a valid reference is ignored, so a
 * malformed tool value can never smuggle a bogus image into the prompt.
 */
export function attachmentRefsOfValue(value: unknown): ImageRef[] {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return [];
  const raw = (value as Record<string, unknown>)["attachments"];
  if (!Array.isArray(raw)) return [];
  const out: ImageRef[] = [];
  for (const item of raw) {
    const ref = normalizeImageRef(item);
    if (ref !== null) out.push(ref);
  }
  return out;
}

// Usage (message.rs:81-108)
// ---------------------------------------------------------------------------

/** Provider-reported token usage for one LLM response (`Usage`). */
export interface Usage {
  prompt_tokens: number;
  completion_tokens: number;
  total_tokens: number;
  cache_read: number;
  reasoning_tokens: number;
}

/** All-zero usage (`Usage::default`). */
export function zeroUsage(): Usage {
  return { prompt_tokens: 0, completion_tokens: 0, total_tokens: 0, cache_read: 0, reasoning_tokens: 0 };
}

/** `Usage::is_empty` — true when every counter is zero. */
export function usageIsEmpty(u: Usage): boolean {
  return (
    u.prompt_tokens === 0 &&
    u.completion_tokens === 0 &&
    u.total_tokens === 0 &&
    u.cache_read === 0 &&
    u.reasoning_tokens === 0
  );
}

/** `Usage::add` — per-field sum, returning a new value. */
export function usageAdd(a: Usage, b: Usage): Usage {
  return {
    prompt_tokens: a.prompt_tokens + b.prompt_tokens,
    completion_tokens: a.completion_tokens + b.completion_tokens,
    total_tokens: a.total_tokens + b.total_tokens,
    cache_read: a.cache_read + b.cache_read,
    reasoning_tokens: a.reasoning_tokens + b.reasoning_tokens,
  };
}

/** Sum a list of usages (`Usage +=` in a loop). */
export function usageSum(usages: readonly Usage[]): Usage {
  let total = zeroUsage();
  for (const u of usages) total = usageAdd(total, u);
  return total;
}

/** Cache-hit ratio as reported by the statusline (`cache_read / prompt_tokens`). */
export function cacheHitRatio(u: Usage): number {
  return u.prompt_tokens === 0 ? 0 : u.cache_read / u.prompt_tokens;
}
